[Wang Ruijie] iP

wrjgold opened the pull request:
# Tick

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))



Tick frees your mind of having to remember things you need to do. It's,


- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

``` Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

wrjgold committed:
Rename bot

wrjgold committed:
Add greeting and reformat to OOP

wrjgold committed:
Rename greet function

wrjgold committed:
Add echo function and exits when user says bye

wrjgold committed:
implement add and list functions

wrjgold committed:
Merge echo and addToList, Change to using ArrayList

wrjgold committed:
Add and implement Task class

wrjgold committed:
Implement marking/unmarking of tasks. Integrate Task class. Refactor code

wrjgold committed:
Implement new types of Task: ToDo, Deadline, Event

wrjgold committed:
Handle 3 new task types

wrjgold committed:
Fix bug where Scanner throws NoSuchElementException

wrjgold committed:
Add test cases

wrjgold committed:
Add exception handling

wrjgold committed:
Modify exception handling to be more general

wrjgold committed:
Add function to delete task

wrjgold committed:
Modify dialogue to add personality

wrjgold committed:
Add enum class for list of available commands

wrjgold committed:
Update test case

wrjgold committed:
Add storage class to save results

wrjgold committed:
Update gitignore for data folder

wrjgold committed:
Merge branch 'branch-Level-7'

wrjgold committed:
Modify tasks to store date time

wrjgold committed:
Merge branch 'branch-Level-8'

wrjgold committed:
Add Ui class

wrjgold committed:
Add Parser class

wrjgold committed:
Add UI class

wrjgold committed:
Add TaskList class

wrjgold committed:
Add various Command classes for abstraction

wrjgold committed:
Add IncorrectCommand class

wrjgold committed:
Modify error message

wrjgold committed:
Modify to utilise more OOP

wrjgold committed:
Merge branch 'moreOOP'

wrjgold committed:
Organize classes into packages

wrjgold committed:
Merge branch 'add-gradle-support'

wrjgold committed:
Configure gradle

wrjgold committed:
Fix wrong text output

wrjgold committed:
Move TaskList to storage package

wrjgold committed:
Add JUnit test cases for TaskList and Parser

wrjgold committed:
Rename output jar file

wrjgold committed:
Add JavaDocs

wrjgold committed:
Merge branch 'branch-A-JavaDoc'

wrjgold committed:
Add Find feature

wrjgold committed:
Change wildcard import to single imports

wrjgold committed:
Merge branch 'branch-A-CodingStandard'

wrjgold committed:
Merge branch 'branch-Level-9'

wrjgold committed:
Fix missing import

wrjgold committed:
Add error handling for index out of bound

wrjgold committed:
Add checkstyle plugin

wrjgold committed:
Fix code style according to checkstyle feedback

wrjgold committed:
Merge branch 'branch-A-CheckStyle'

CJianzhi commented on src/main/java/tick/commands/AddCommand.java:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * Returns false as the command is not an exit command.
+     *
+     * @return false
```
Should there be a full stop added after false?

CJianzhi commented on src/main/java/tick/parser/Parser.java:
```diff
@@ -0,0 +1,110 @@
+            default:
+                return new IncorrectCommand();
+            }
+        } catch (IllegalArgumentException e) {
```
I really like how you separated the two Exceptions here and displayed different messages!

CJianzhi commented on src/main/java/tick/storage/Storage.java:
```diff
@@ -0,0 +1,96 @@
+            String description = parts[2];
+            Task task;
+
+            switch (taskType) {
```
I like the use of case clauses here as well!

SuhailLoya replied:
I agree, the code is very neat and clean

CJianzhi commented on src/main/java/tick/tasks/Event.java:
```diff
@@ -0,0 +1,48 @@
+     * @return A string representation of the event task for file storage.
+     */
+    public String toFileString() {
+        return String.format("E | %s | %s | %s | %s", super.getStatus() ? "1" : "0", super.getDescription(),
```
Should this line be wrapped at the comma after "0"?

CJianzhi commented on src/main/java/tick/parser/Parser.java:
```diff
@@ -0,0 +1,110 @@
+import java.time.DateTimeException;
+import java.time.LocalDate;
+
+import tick.commands.AddCommand;
```
I like how imports are arranged alphabetically here and no wildcard imports used!

CJianzhi reviewed (commented):
LGTM. Coding standards are adhered to with no major violations.

wrjgold committed:
Add GUI functionality

wrjgold committed:
Merge branch 'branch-Level-10'

wrjgold committed:
Fix style for newly added GUI classes

SuhailLoya commented on src/main/java/tick/Tick.java:
```diff
@@ -0,0 +1,87 @@
+ * Tick is a chatbot that helps users to keep track of their tasks.
+ */
+public class Tick {
+    private TaskList tasks;
+    private Storage storage;
+    private Ui ui;
```
I like the clear and descriptive variable names

SuhailLoya commented on src/main/java/tick/Tick.java:
```diff
@@ -0,0 +1,87 @@
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                ui.showLine(); // show the divider line ("_______")
+                Command c = Parser.parse(fullCommand);
+                c.execute(tasks, ui, storage);
+                isExit = c.isExit();
+            } catch (TickException e) {
+                ui.showError(e.getMessage());
+            } finally {
+                ui.showLine();
+            }
```
I like the way commands are handled here, very neat!

SuhailLoya commented on src/main/java/tick/commands/AddCommand.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class AddCommand extends Command {
+    private Task toAdd;
+    private CommandType type = CommandType.ADD;
```
I am not sure if this works? Perhaps you can leverage the `type` variable at the Command class and use a setter (or alternatively add it to the constructor of Command class). Same goes for other commands :)

SuhailLoya commented on src/main/java/tick/commands/CommandType.java:
```diff
@@ -0,0 +1,17 @@
+    EVENT,
+    FIND,
+    ADD
+}
```
I like the use of enum

SuhailLoya commented on src/main/java/tick/parser/Parser.java:
```diff
@@ -0,0 +1,110 @@
+        return new AddCommand(task);
+    }
+
+    private static Command prepareDeadline(String args) throws TickException {
```
I like the use of verbs as method names. Perhaps a more descriptive and common method name would be `handleDeadline`

SuhailLoya commented on src/main/java/tick/storage/Storage.java:
```diff
@@ -0,0 +1,96 @@
+            default:
+                continue;
+            }
+            if (isDone) {
+                task.markAsDone();
+            }
```
I like the use of markAsDone() here, this way you only have one place handling the logic of markAsDone!

SuhailLoya commented on src/main/java/tick/storage/TaskList.java:
```diff
@@ -0,0 +1,137 @@
+ * Represents a list of tasks.
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
```
I like that you used plural name here `tasks`. It shows that this is a collection

SuhailLoya commented on src/main/java/tick/storage/TaskList.java:
```diff
@@ -0,0 +1,137 @@
+/**
+ * Represents a list of tasks.
+ */
+public class TaskList {
```
I like the descriptive javadocs and method names. Method names are verbs and clearly describe what they are supposed to do

SuhailLoya commented on src/main/java/tick/storage/TaskList.java:
```diff
@@ -0,0 +1,137 @@
+        }
+    }
+
+    /**
+     * Gets all tasks in the TaskList.
+     *
+     * @return ArrayList of tasks in the TaskList.
+     */
```
I like that you provided java docs here even though they are optional for getters / setters :)

SuhailLoya commented on src/main/java/tick/tasks/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+ * Represents a deadline task which has a deadline.
+ */
+public class Deadline extends Task {
+    private LocalDate by;
```
perhaps the naming of the variable by can be more descriptive. Suggestions: `dueBy` `byDate`. Same goes date variables at other places :)

SuhailLoya commented on src/main/java/tick/tasks/Task.java:
```diff
@@ -0,0 +1,78 @@
+ */
+public abstract class Task {
+    private String description;
+    private boolean isDone;
```
i like the variable name `isDone`. Clearly describes it is a boolean variable representing if the task is done

SuhailLoya commented on src/main/java/tick/ui/Ui.java:
```diff
@@ -0,0 +1,173 @@
+/**
+ * Ui class handles the user interface of the application.
+ */
+public class Ui {
```
very readable :)

SuhailLoya commented on src/main/java/tick/tasks/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+     *
+     * @return A string representation of the deadline task for file storage.
+     */
+    public String toFileString() {
```
Should consider adding `@Override` annotation. Same goes for other overriding methods too :)

SuhailLoya commented on src/main/java/tick/tasks/Task.java:
```diff
@@ -0,0 +1,78 @@
+     *
+     * @return A string representation of the task that can be saved to a file.
+     */
+    public abstract String toFileString();
```
the name `toFileString` might cause some confusion. Perhaps a more descriptive name would be `toStorageFormat`

SuhailLoya reviewed (commented):
LGTM!! Just a few nits to fix

wrjgold committed:
Remove unused command types

Command classes: Remove unused CommandType field

CommandType field is not being used at allX

wrjgold committed:
Remove unused command type

CommandType Class: Remove ADD command type
ADD command type is unused throughout the program

wrjgold committed:
Add fullstop to javadocs

wrjgold committed:
Rename method that converts tasks to file storage format

wrjgold committed:
Rename date variables

wrjgold committed:
Add varargs constructor for TaskList

wrjgold committed:
Add Java assertions

There are several invariants that should be verified in the program.
Added java assertions to verfiy these assumptions.

This helps to verify assumptions made throughout the program and catch
errors early, adding as an extra layer of error handling.

Let's
* Add assertions to check that Task or TaskList are not null values
* Add assertions to check that arguments for Deadline and Event tasks
  are correct

wrjgold committed:
Refactor code and improve code quality

Improve readability and quality of code.

Let's
* Implement default branch for switch statement to detect error in file
  formatting
* Restructure code to improve readability in Storage class

wrjgold committed:
Implement find task using streams

Find tasks with matching keywords can be implemented with the help of
stream functions.

wrjgold committed:
Merge pull request #1 from wrjgold/branch-A-Assertions

Add Java assertions

wrjgold committed:
Merge branch 'master' into branch-A-CodeQuality

wrjgold committed:
Merge pull request #2 from wrjgold/branch-A-CodeQuality

Refactor code and improve code quality

wrjgold committed:
Merge branch 'master' into branch-A-Streams

wrjgold committed:
Merge pull request #3 from wrjgold/branch-A-Streams

Implement find task using streams

wrjgold committed:
Setup github actions for CI/CD

wrjgold committed:
Implement duplication check when adding tasks

There is a possibility where users add repeated tasks with the same
description. This check is to prevent users from accidentally adding
duplicate tasks.

Let's
* Add duplication check in TaskList
* Check if tasks are equal based on their description

wrjgold committed:
Merge branch 'branch-BCD-Extension'

wrjgold committed:
Tweak GUI to polish interface

wrjgold committed:
Add personality to Tick

wrjgold committed:
Add tests for DeleteCommand

wrjgold committed:
Modify test to use JUnit methods

wrjgold committed:
Add initial greeting to GUI

wrjgold committed:
Update user guide

wrjgold committed:
Add more unit tests

wrjgold committed:
Modify Readme.md file

wrjgold committed:
Fix wrongly named test package

wrjgold committed:
Fix DeleteCommand JUnit test

wrjgold committed:
Fix checkstyle error
